tensorflow_addons/callbacks/tests/avg_model_checkpoint_test.py

```diff
@@ -0,0 +1,206 @@
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    model.compile(loss="categorical_crossentropy", optimizer=optimizer, metrics=["acc"])
+    return x, y, model
+
+
+def test_compatibility_with_some_opts_only():
+    x, y, model = get_data_and_model(optimizer="rmsprop")
+    avg_model_ckpt = AverageModelCheckpoint(
+        update_weights=True, filepath=test_model_filepath
+    )
+    with pytest.raises(
+        TypeError,
+        match="AverageModelCheckpoint is only used when trainingwith"
+        + " MovingAverage or StochasticAverage",
+    ):
+        assert model.fit(
+            x, y, epochs=EPOCHS, batch_size=BATCH_SIZE, callbacks=[avg_model_ckpt]
+        )
+
+
+def test_model_file_creation():
+    x, y, model = get_data_and_model()
+    avg_model_ckpt = AverageModelCheckpoint(
+        update_weights=True, filepath=test_model_filepath
+    )
+    model.fit(x, y, epochs=EPOCHS, batch_size=BATCH_SIZE, callbacks=[avg_model_ckpt])
+    assert os.path.exists(test_model_filepath)
+    os.remove(test_model_filepath)
+
+
+def test_mode_auto():
+    x, y, model = get_data_and_model()
+    monitor = "val_loss"
+    save_best_only = False
+    mode = "auto"
+    avg_model_ckpt = AverageModelCheckpoint(
+        update_weights=True,
+        filepath=test_model_filepath,
+        monitor=monitor,
+        save_best_only=save_best_only,
+        mode=mode,
+    )
+    model.fit(
+        x,
+        y,
+        epochs=EPOCHS,
+        batch_size=BATCH_SIZE,
+        validation_data=(x, y),
+        callbacks=[avg_model_ckpt],
+    )
+    assert os.path.exists(test_model_filepath)
+    os.remove(test_model_filepath)
+
+
+def test_mode_min():
+    x, y, model = get_data_and_model()
+    monitor = "val_loss"
+    save_best_only = False
+    mode = "min"
+    avg_model_ckpt = AverageModelCheckpoint(
+        update_weights=True,
+        filepath=test_model_filepath,
+        monitor=monitor,
+        save_best_only=save_best_only,
+        mode=mode,
+    )
+    model.fit(
+        x,
+        y,
+        epochs=EPOCHS,
+        batch_size=BATCH_SIZE,
+        validation_data=(x, y),
+        callbacks=[avg_model_ckpt],
+    )
+    assert os.path.exists(test_model_filepath)
+    os.remove(test_model_filepath)
+
+
+def test_mode_max():
+    x, y, model = get_data_and_model()
+    mode = "max"
+    monitor = "val_acc"
+    save_best_only = False
+    avg_model_ckpt = AverageModelCheckpoint(
+        update_weights=True,
+        filepath=test_model_filepath,
+        monitor=monitor,
+        save_best_only=save_best_only,
+        mode=mode,
+    )
+    model.fit(
+        x,
+        y,
+        epochs=EPOCHS,
+        batch_size=BATCH_SIZE,
+        validation_data=(x, y),
+        callbacks=[avg_model_ckpt],
+    )
+    assert os.path.exists(test_model_filepath)
+    os.remove(test_model_filepath)
+
+
+def test_save_best_only():
+    x, y, model = get_data_and_model()
+    save_best_only = True
```

Comment:
Here we test that the callback runs, which is good, but we could also check for correctness (the callback does what we want it to do). 

To do that, grab the keras history in the fit, load the saved model, do an evaluation pass and ensure the metric is equal to the best one found in the keras history.


I don't know if it's possible to do that for other tests, but if it's possible, we should aim for that.

Reply:
@gabrieldemarmiesse , please refer to [this comment](https://github.com/keras-team/keras/issues/6977#issuecomment-400207482). I don't think it is a good idea to test `fit()` results with `evaluate()`

Reply:
You're right, you'll have problems if you compare on the train data. On the validation data, the keras history should give you the same result as a standalone evaluate call since the model does't change during an evaluation pass.

Reply:
FYI, all types of data here are same. And I have tested it and after testing it, I can't strongly suggest that method because the results I was getting are not good.

Reply:
Maybe We should use any particular simple data which doesn't deviate the model.

Reply:
@ashutosh1919 can you paste your code here which you have tested?

Reply:
If it doesn't work, it may be that there is a bug in the implementation of the callback, in which case we should decorate the test with `@pytest.xfail` and open an issue to track the bug. @ashutosh1919 we won't ask you to fix the bug, it's outside the scope of this pull request.

Reply:
@gabrieldemarmiesse , I think I got the idea of testing this. We need to test the model weights. Got the idea from some tests in tf repo.

Reply:
I would have expected that weird behavior in earlier versions because of how batchnorm used to operate but that has been resolved long ago. So, IMO, `evaluate` would give you the same result as during the `fit`